src/Elastic.Apm/DistributedTracing/TraceParent.cs

```diff
@@ -0,0 +1,199 @@
+		/// <param name="prentId">The parsed parentId</param>
+		/// <param name="traceOptions">The parsed traceOptions flags</param>
+		/// <returns>True if parsing was successful, false otherwise.</returns>
+		internal static bool TryExtractTraceparent(string traceParentValue, out string traceId, out string prentId, out byte[] traceOptions)
+		{
+			traceId = string.Empty;
+			prentId = string.Empty;
+			traceOptions = null;
+
+			var bestAttempt = false;
+
+			if (string.IsNullOrWhiteSpace(traceParentValue)) return false;
+
+			if (traceParentValue.Length < VersionPrefixIdLength || traceParentValue[VersionPrefixIdLength - 1] != '-') return false;
+
+			try
+			{
+				var versionArray = StringToByteArray(traceParentValue, 0, VersionLength);
+				if (versionArray[0] == 255)
+					return false;
+
+				if (versionArray[0] > 0)
+					// expected version is 00
+					// for higher versions - best attempt parsing of trace id, span id, etc.
+					bestAttempt = true;
+			}
+			catch
+			{
+				return false;
+			}
+
+			if (traceParentValue.Length < VersionAndTraceIdLength || traceParentValue[VersionAndTraceIdLength - 1] != '-') return false;
+
+			try
+			{
+				//parse value to byte array and back to string:
+				//TODO: probably we should just validate the chars
+				var traceIdVal = ByteArrayToString(StringToByteArray(traceParentValue, VersionPrefixIdLength, TraceIdLength));
+
+				if (traceIdVal == "00000000000000000000000000000000")
+					return false;
+
+				traceId = traceIdVal;
+			}
+			catch (ArgumentOutOfRangeException)
+			{
+				return false;
+			}
+
+			if (traceParentValue.Length < VersionAndTraceIdAndSpanIdLength
+				|| traceParentValue[VersionAndTraceIdAndSpanIdLength - 1] != '-') return false;
+
+			try
+			{
+				//parse value to byte array and back to string:
+				//TODO: probably we should just validate the chars
+				var prentIdVal = ByteArrayToString(StringToByteArray(traceParentValue, VersionAndTraceIdLength, SpanIdLength));
+
+				if (prentIdVal == "0000000000000000")
+					return false;
+
+				prentId = prentIdVal;
+			}
+			catch (ArgumentOutOfRangeException)
+			{
+				return false;
+			}
+
+			if (traceParentValue.Length < VersionAndTraceIdAndSpanIdLength + OptionsLength) return false;
+
+			try
+			{
+				traceOptions = StringToByteArray(traceParentValue, VersionAndTraceIdAndSpanIdLength, OptionsLength);
+			}
+			catch (ArgumentOutOfRangeException)
+			{
+				return false;
+			}
+
+			if (!bestAttempt && traceParentValue.Length != VersionAndTraceIdAndSpanIdLength + OptionsLength) return false;
+
+			// ReSharper disable once InvertIf - imo that would make it hard to read.
+			if (bestAttempt)
+			{
+				if (traceParentValue.Length > VersionAndTraceIdAndSpanIdLength + OptionsLength &&
+					traceParentValue[VersionAndTraceIdAndSpanIdLength + OptionsLength] != '-')
+					return false;
+			}
+
+			return true;
+		}
+
+		private static string ByteArrayToString(IEnumerable<byte> bytes)
```

Comment:
That is a lot of code (`ByteArrayToString` and its dependencies) to just verify that string contains only hex digits. Do you think it's worth?

Reply:
No, that was related to my perf. todo. 
That method got removed:  https://github.com/elastic/apm-agent-dotnet/pull/168#issuecomment-481003766